diff --git a/chrome/browser/extensions/api/settings_private/prefs_util.cc b/chrome/browser/extensions/api/settings_private/prefs_util.cc
index e81d7f44bfd9f..f05f3ec469b1a 100644
--- a/chrome/browser/extensions/api/settings_private/prefs_util.cc
+++ b/chrome/browser/extensions/api/settings_private/prefs_util.cc
@@ -403,6 +403,7 @@ const PrefsUtil::TypedPrefMap& PrefsUtil::GetAllowlistedKeys() {
   // Privacy Guide
   (*s_allowlist)[::prefs::kPrivacyGuideViewed] =
       settings_api::PrefType::kBoolean;
+  (*s_allowlist)[::prefs::kEnableGPC] = settings_api::PrefType::kBoolean;
 
   // Privacy Sandbox page
   (*s_allowlist)[::prefs::kPrivacySandboxM1TopicsEnabled] =
diff --git a/chrome/browser/renderer_preferences_util.cc b/chrome/browser/renderer_preferences_util.cc
index 254a3b5a0cc92..586d6fc4e98b6 100644
--- a/chrome/browser/renderer_preferences_util.cc
+++ b/chrome/browser/renderer_preferences_util.cc
@@ -152,6 +152,7 @@ void UpdateFromSystemSettings(blink::RendererPreferences* prefs,
   prefs->enable_referrers = pref_service->GetBoolean(prefs::kEnableReferrers);
   prefs->enable_do_not_track =
       pref_service->GetBoolean(prefs::kEnableDoNotTrack);
+  prefs->enable_gpc = pref_service->GetBoolean(prefs::kEnableGPC);
   prefs->enable_encrypted_media =
       pref_service->GetBoolean(prefs::kEnableEncryptedMedia);
 
diff --git a/chrome/browser/ui/browser_ui_prefs.cc b/chrome/browser/ui/browser_ui_prefs.cc
index 970e232178194..325ec109e759f 100644
--- a/chrome/browser/ui/browser_ui_prefs.cc
+++ b/chrome/browser/ui/browser_ui_prefs.cc
@@ -145,6 +145,9 @@ void RegisterBrowserUserPrefs(user_prefs::PrefRegistrySyncable* registry) {
   registry->RegisterBooleanPref(
       prefs::kEnableDoNotTrack, false,
       user_prefs::PrefRegistrySyncable::SYNCABLE_PREF);
+  registry->RegisterBooleanPref(
+      prefs::kEnableGPC, true,
+      user_prefs::PrefRegistrySyncable::SYNCABLE_PREF);
 #if !BUILDFLAG(IS_CHROMEOS) && !BUILDFLAG(IS_ANDROID)
   registry->RegisterBooleanPref(prefs::kPrintPreviewUseSystemDefaultPrinter,
                                 false);
diff --git a/chrome/browser/ui/prefs/pref_watcher.cc b/chrome/browser/ui/prefs/pref_watcher.cc
index 4cae431a3012e..81fbdafb13566 100644
--- a/chrome/browser/ui/prefs/pref_watcher.cc
+++ b/chrome/browser/ui/prefs/pref_watcher.cc
@@ -152,6 +152,7 @@ PrefWatcher::PrefWatcher(Profile* profile) : profile_(profile) {
                                      renderer_callback);
   profile_pref_change_registrar_.Add(prefs::kEnableDoNotTrack,
                                      renderer_callback);
+  profile_pref_change_registrar_.Add(prefs::kEnableGPC, renderer_callback);
 
 #if !BUILDFLAG(IS_MAC)
   profile_pref_change_registrar_.Add(prefs::kFullscreenAllowed,
diff --git a/chrome/common/pref_names.h b/chrome/common/pref_names.h
index 08acbb0628037..a84b48742a3f9 100644
--- a/chrome/common/pref_names.h
+++ b/chrome/common/pref_names.h
@@ -1010,6 +1010,10 @@ inline constexpr char kEnableReferrers[] = "enable_referrers";
 // Whether to send the DNT header.
 inline constexpr char kEnableDoNotTrack[] = "enable_do_not_track";
 
+// Whether to send the Global Privacy Control header and expose
+// navigator.globalPrivacyControl.
+inline constexpr char kEnableGPC[] = "enable_gpc";
+
 // Whether to allow the use of Encrypted Media Extensions (EME), except for the
 // use of Clear Key key sytems, which is always allowed as required by the spec.
 // TODO(crbug.com/40549758): This pref was used as a WebPreference which is why
diff --git a/content/browser/loader/browser_initiated_resource_request.cc b/content/browser/loader/browser_initiated_resource_request.cc
index a8cdb42779b98..36efd26e75ca8 100644
--- a/content/browser/loader/browser_initiated_resource_request.cc
+++ b/content/browser/loader/browser_initiated_resource_request.cc
@@ -30,6 +30,12 @@ void UpdateAdditionalHeadersForBrowserInitiatedRequest(
     }
     headers->SetHeaderIfMissing(blink::kDoNotTrackHeader, "1");
   }
+  if (renderer_preferences.enable_gpc) {
+    if (should_update_existing_headers) {
+      headers->RemoveHeader(blink::kSecGPCHeader);
+    }
+    headers->SetHeaderIfMissing(blink::kSecGPCHeader, "1");
+  }
 
   // TODO(crbug.com/40833603): WARNING: This bypasses the permissions policy.
   // Unfortunately, workers lack a permissions policy and to derive proper hints
diff --git a/content/renderer/render_frame_impl.cc b/content/renderer/render_frame_impl.cc
index 16b998336897f..b250097276525 100644
--- a/content/renderer/render_frame_impl.cc
+++ b/content/renderer/render_frame_impl.cc
@@ -4514,6 +4518,10 @@ void RenderFrameImpl::FinalizeRequestInternal(
     request.SetHttpHeaderField(
         blink::WebString::FromUtf8(blink::kDoNotTrackHeader), "1");
   }
+  if (GetWebView()->GetRendererPreferences().enable_gpc) {
+    request.SetHttpHeaderField(
+        blink::WebString::FromUtf8(blink::kSecGPCHeader), "1");
+  }
 
   // The request's extra data may indicate that we should set a custom user
   // agent. This needs to be done here, after WebKit is through with setting the
diff --git a/third_party/blink/common/loader/loader_constants.cc b/third_party/blink/common/loader/loader_constants.cc
index ae82ae372f79b..beb3409c46d79 100644
--- a/third_party/blink/common/loader/loader_constants.cc
+++ b/third_party/blink/common/loader/loader_constants.cc
@@ -7,5 +7,6 @@
 namespace blink {
 
 const char kDoNotTrackHeader[] = "DNT";
+const char kSecGPCHeader[] = "Sec-GPC";
 
 }  // namespace blink
diff --git a/third_party/blink/common/renderer_preferences/renderer_preferences_mojom_traits.cc b/third_party/blink/common/renderer_preferences/renderer_preferences_mojom_traits.cc
index a42179970119c..a388a68616def 100644
--- a/third_party/blink/common/renderer_preferences/renderer_preferences_mojom_traits.cc
+++ b/third_party/blink/common/renderer_preferences/renderer_preferences_mojom_traits.cc
@@ -54,6 +54,7 @@ bool StructTraits<blink::mojom::RendererPreferencesDataView,
   out->enable_referrers = data.enable_referrers();
   out->allow_cross_origin_auth_prompt = data.allow_cross_origin_auth_prompt();
   out->enable_do_not_track = data.enable_do_not_track();
+  out->enable_gpc = data.enable_gpc();
   out->enable_encrypted_media = data.enable_encrypted_media();
 
   if (!data.ReadWebrtcIpHandlingPolicy(&out->webrtc_ip_handling_policy))
diff --git a/third_party/blink/public/common/loader/loader_constants.h b/third_party/blink/public/common/loader/loader_constants.h
index 7b3fb4f53f734..1fb38d32e4863 100644
--- a/third_party/blink/public/common/loader/loader_constants.h
+++ b/third_party/blink/public/common/loader/loader_constants.h
@@ -11,6 +11,7 @@ namespace blink {
 
 // HTTP header set in requests to indicate they should be marked DoNotTrack.
 BLINK_COMMON_EXPORT extern const char kDoNotTrackHeader[];
+BLINK_COMMON_EXPORT extern const char kSecGPCHeader[];
 
 // These values indicate the load progress constants shared between both
 // //content and //blink.
diff --git a/third_party/blink/public/common/renderer_preferences/renderer_preferences.h b/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
index 08d068cc0af4e..45d4bb340901a 100644
--- a/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
+++ b/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
@@ -52,6 +52,7 @@ struct BLINK_COMMON_EXPORT RendererPreferences {
   bool enable_referrers{true};
   bool allow_cross_origin_auth_prompt{false};
   bool enable_do_not_track{false};
+  bool enable_gpc{false};
   bool enable_encrypted_media{true};
 #if BUILDFLAG(IS_CHROMEOS)
   bool use_overlay_scrollbar{false};
diff --git a/third_party/blink/public/common/renderer_preferences/renderer_preferences_mojom_traits.h b/third_party/blink/public/common/renderer_preferences/renderer_preferences_mojom_traits.h
index 11270bb1e2293..e25d101022069 100644
--- a/third_party/blink/public/common/renderer_preferences/renderer_preferences_mojom_traits.h
+++ b/third_party/blink/public/common/renderer_preferences/renderer_preferences_mojom_traits.h
@@ -128,6 +128,10 @@ struct BLINK_COMMON_EXPORT
     return data.enable_do_not_track;
   }
 
+  static const bool& enable_gpc(const ::blink::RendererPreferences& data) {
+    return data.enable_gpc;
+  }
+
   static const bool& enable_encrypted_media(
       const ::blink::RendererPreferences& data) {
     return data.enable_encrypted_media;
diff --git a/third_party/blink/public/mojom/renderer_preferences.mojom b/third_party/blink/public/mojom/renderer_preferences.mojom
index aa6d396335d33..d77218f6cfffe 100644
--- a/third_party/blink/public/mojom/renderer_preferences.mojom
+++ b/third_party/blink/public/mojom/renderer_preferences.mojom
@@ -89,6 +89,9 @@ struct RendererPreferences {
   // Set to true to indicate that the preference to set DNT to 1 is enabled.
   bool enable_do_not_track = false;
 
+  // Set to true to indicate that the preference to set Sec-GPC to 1 is enabled.
+  bool enable_gpc = false;
+
   // Whether to allow the use of Encrypted Media Extensions (EME), except for
   // the use of Clear Key key system which is always allowed as required by the
   // spec.
diff --git a/third_party/blink/renderer/bindings/idl_in_modules.gni b/third_party/blink/renderer/bindings/idl_in_modules.gni
index d968cedda43df..c49e19aa1da54 100644
--- a/third_party/blink/renderer/bindings/idl_in_modules.gni
+++ b/third_party/blink/renderer/bindings/idl_in_modules.gni
@@ -1293,6 +1293,11 @@ static_idl_files_in_modules = [
   "//third_party/blink/renderer/modules/xr/xr_webgl_sub_image.idl",
 ]
 
+static_idl_files_in_modules += [
+  "//third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.idl",
+  "//third_party/blink/renderer/modules/global_privacy_control/worker_navigator_global_privacy_control.idl",
+]
+
 # Direct Sockets
 # This uses target_os rather than current_os (which is what is_android is set
 # from) for the case of generating the v8 context snapshot for android. When
diff --git a/third_party/blink/renderer/core/exported/web_view_impl.cc b/third_party/blink/renderer/core/exported/web_view_impl.cc
index 92c8a39a2572a..996b31f05408b 100644
--- a/third_party/blink/renderer/core/exported/web_view_impl.cc
+++ b/third_party/blink/renderer/core/exported/web_view_impl.cc
@@ -3618,6 +3618,9 @@ void WebViewImpl::UpdateRendererPreferences(
   GetSettings()->SetCaretBrowsingEnabled(
       renderer_preferences_.caret_browsing_enabled);
 
+  WebRuntimeFeatures::EnableGlobalPrivacyControl(
+      renderer_preferences_.enable_gpc);
+
 #if BUILDFLAG(IS_OZONE)
   GetSettings()->SetSelectionClipboardBufferAvailable(
       renderer_preferences_.selection_clipboard_buffer_available);
diff --git a/third_party/blink/renderer/modules/BUILD.gn b/third_party/blink/renderer/modules/BUILD.gn
index 842b7834acd9e..708cc9e6226a5 100644
--- a/third_party/blink/renderer/modules/BUILD.gn
+++ b/third_party/blink/renderer/modules/BUILD.gn
@@ -100,6 +100,7 @@ component("modules") {
     "//third_party/blink/renderer/modules/file_system_access",
     "//third_party/blink/renderer/modules/font_access",
     "//third_party/blink/renderer/modules/gamepad",
+    "//third_party/blink/renderer/modules/global_privacy_control",
     "//third_party/blink/renderer/modules/handwriting",
     "//third_party/blink/renderer/modules/hid",
     "//third_party/blink/renderer/modules/idle",
diff --git a/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc b/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc
index 080b9b445198b..44644931c6158 100644
--- a/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc
+++ b/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc
@@ -150,6 +150,9 @@ void WebServiceWorkerFetchContextImpl::FinalizeRequest(WebURLRequest& request) {
   if (renderer_preferences_.enable_do_not_track) {
     request.SetHttpHeaderField(WebString::FromUtf8(kDoNotTrackHeader), "1");
   }
+  if (renderer_preferences_.enable_gpc) {
+    request.SetHttpHeaderField(WebString::FromUtf8(kSecGPCHeader), "1");
+  }
   auto url_request_extra_data = base::MakeRefCounted<WebURLRequestExtraData>();
   url_request_extra_data->set_originated_from_service_worker(true);
 
diff --git a/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc b/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc
index e0e9af90a35fd..8370aae45dd53 100644
--- a/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc
+++ b/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc
@@ -377,6 +377,9 @@ void DedicatedOrSharedWorkerGlobalScopeContextImpl::FinalizeRequest(
   if (renderer_preferences_.enable_do_not_track) {
     request.SetHttpHeaderField(WebString::FromUtf8(kDoNotTrackHeader), "1");
   }
+  if (renderer_preferences_.enable_gpc) {
+    request.SetHttpHeaderField(WebString::FromUtf8(kSecGPCHeader), "1");
+  }
 
   auto url_request_extra_data = base::MakeRefCounted<WebURLRequestExtraData>();
   request.SetURLRequestExtraData(std::move(url_request_extra_data));
diff --git a/third_party/blink/renderer/platform/runtime_enabled_features.json5 b/third_party/blink/renderer/platform/runtime_enabled_features.json5
index 1b59fe994a8b1..2eef5e2b29abf 100644
--- a/third_party/blink/renderer/platform/runtime_enabled_features.json5
+++ b/third_party/blink/renderer/platform/runtime_enabled_features.json5
@@ -3175,6 +3175,11 @@
       name: "GetUserMediaEchoCancellationModes",
       status: "stable",
     },
+    {
+      name: "GlobalPrivacyControl",
+      base_feature_status: "disabled",
+      public: true,
+    },
     {
       name: "GroupEffect",
       status: "test",
