diff --git a/chrome/browser/ui/webui/BUILD.gn b/chrome/browser/ui/webui/BUILD.gn
index fa8bdb024fbee..517347e91bf66 100644
--- a/chrome/browser/ui/webui/BUILD.gn
+++ b/chrome/browser/ui/webui/BUILD.gn
@@ -32,6 +32,7 @@ source_set("configs") {
     "chrome_untrusted_web_ui_configs.h",
     "chrome_web_ui_configs.cc",
     "chrome_web_ui_configs.h",
+    "thorium_easter_eggs.h",
   ]
 
   configs += [ "//build/config/compiler:wexit_time_destructors" ]
diff --git a/chrome/browser/ui/webui/chrome_web_ui_configs.cc b/chrome/browser/ui/webui/chrome_web_ui_configs.cc
index 9daef52054236..d50f7f9b10295 100644
--- a/chrome/browser/ui/webui/chrome_web_ui_configs.cc
+++ b/chrome/browser/ui/webui/chrome_web_ui_configs.cc
@@ -54,6 +54,7 @@
 #include "chrome/browser/ui/webui/signin_internals_ui.h"
 #include "chrome/browser/ui/webui/subresource_filter/subresource_filter_internals_ui.h"
 #include "chrome/browser/ui/webui/sync_internals/sync_internals_ui.h"
+#include "chrome/browser/ui/webui/thorium_easter_eggs.h"
 #include "chrome/browser/ui/webui/translate_internals/translate_internals_ui.h"
 #include "chrome/browser/ui/webui/usb_internals/usb_internals_ui.h"
 #include "chrome/browser/ui/webui/version/version_ui.h"
@@ -297,6 +298,7 @@ void RegisterChromeWebUIConfigs() {
                      subresource_filter::SubresourceFilterInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<SiteEngagementUIConfig>());
   map.AddWebUIConfig(std::make_unique<SyncInternalsUIConfig>());
+  map.AddWebUIConfig(std::make_unique<ThoriumWebUILoadUIConfig>());
   map.AddWebUIConfig(std::make_unique<TranslateInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<UsbInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<user_actions_ui::UserActionsUIConfig>());
diff --git a/chrome/browser/ui/webui/chrome_web_ui_controller_factory.cc b/chrome/browser/ui/webui/chrome_web_ui_controller_factory.cc
index ec24b0ffaf339..1dfba3b186b6e 100644
--- a/chrome/browser/ui/webui/chrome_web_ui_controller_factory.cc
+++ b/chrome/browser/ui/webui/chrome_web_ui_controller_factory.cc
@@ -352,7 +352,8 @@ bool ChromeWebUIControllerFactory::IsWebUIAllowedToMakeNetworkRequests(
       // https://crbug.com/40583261
       origin.host() == chrome::kChromeUIDownloadsHost ||
       // https://crbug.com/376417346
-      origin.host() == chrome::kChromeUIExtensionsHost;
+      origin.host() == chrome::kChromeUIExtensionsHost ||
+      origin.host() == chrome::kChromeUIEggsHost;
 }
 
 ChromeWebUIControllerFactory::ChromeWebUIControllerFactory() = default;
diff --git a/chrome/common/webui_url_constants.cc b/chrome/common/webui_url_constants.cc
index dd37ab85cec11..01e6bd243b48c 100644
--- a/chrome/common/webui_url_constants.cc
+++ b/chrome/common/webui_url_constants.cc
@@ -121,6 +121,7 @@ bool IsSystemWebUIHost(std::string_view host) {
 // These hosts will also be suggested by BuiltinProvider.
 base::span<const base::cstring_view> ChromeURLHosts() {
   static constexpr auto kChromeURLHosts = std::to_array<base::cstring_view>({
+      kChromeUIEggsHost,
       kChromeUIAboutHost,
       kChromeUIAccessibilityHost,
       kChromeUIActorInternalsHost,
diff --git a/chrome/common/webui_url_constants.h b/chrome/common/webui_url_constants.h
index e128efef6cf36..d2c1a60747f50 100644
--- a/chrome/common/webui_url_constants.h
+++ b/chrome/common/webui_url_constants.h
@@ -33,6 +33,8 @@ namespace chrome {
 // Not all components have corresponding URLs and vice versa. Only add as
 // needed.
 // Please keep in alphabetical order, with OS/feature specific sections below.
+inline constexpr char kChromeUIEggsHost[] = "eggs";
+inline constexpr char kChromeUIEggsURL[] = "chrome://eggs/";
 inline constexpr char kChromeUIAboutHost[] = "about";
 inline constexpr char kChromeUIAboutURL[] = "chrome://about/";
 inline constexpr char kChromeUIAccessCodeCastHost[] = "access-code-cast";
diff --git a/components/webui/about/resources/about_credits_reciprocal.tmpl b/components/webui/about/resources/about_credits_reciprocal.tmpl
index 5096175e285c1..6d4a41ab2815f 100644
--- a/components/webui/about/resources/about_credits_reciprocal.tmpl
+++ b/components/webui/about/resources/about_credits_reciprocal.tmpl
@@ -1,4 +1,4 @@
 <div class="open-sourced">
-    {{opensource_project}} software is made available as source code
-    <a href="{{opensource_link}}">here</a>.
+    {{thorium_project}} is based on {{opensource_project}} and is made available as source code
+    <a href="{{thorium_link}}">here</a>. Chromium source code is available <a href="{{opensource_link}}">here</a>.
 </div>
diff --git a/ui/webui/resources/BUILD.gn b/ui/webui/resources/BUILD.gn
index ff89ccb3a6deb..9dc03893cf506 100644
--- a/ui/webui/resources/BUILD.gn
+++ b/ui/webui/resources/BUILD.gn
@@ -38,6 +38,7 @@ generate_grd("build_grd") {
     "images:build_grdp",
     "js:build_grdp",
     "mojo:build_grdp",
+    "text:build_grdp",
     "//third_party/lit/v3_0:build_grdp",
   ]
   grdp_files = [
@@ -48,6 +49,7 @@ generate_grd("build_grd") {
     "$target_gen_dir/images/resources.grdp",
     "$target_gen_dir/js/resources.grdp",
     "$target_gen_dir/mojo/resources.grdp",
+    "$target_gen_dir/text/resources.grdp",
   ]
 
   if ((!is_android && !is_ios) || enable_webui_ntp || is_desktop_android) {
